Use standard Redux dispatch to replace most `connectWithActions`.

roberthoenig opened the pull request:
A couple remarks:
* The last commit exists because I forgot to check my commits with prettier. After I had commited everything prettier complained about 7 files. Rebasing felt like too much of a pain, so I hope it's okay this time to have a fix-the-broken-style-of-the-previous-commits commit.

* There were several typing issues popping up. Unfortunately, they all seemed not straightforward to fix, so I'd like to discuss them before digging too deep into them. I'll list them in a new comment.

borisyankov commented:
Are you not using a prettier plugin inside your IDE?

roberthoenig commented:
Some thoughts about type issues that came up:

* In RN, several components have an`onPress()` handler. `onPress()` takes a function f and calls f "on press". I checked all the different onPress docs for RN components, and for all of them the return value of f doesn't play any role. Yet, we have a discrepancy in our own type definitions about that return type: In one place, it's `  onPress: () => void`, in another `onPress: () => void | Promise<any>` and in a third `onPress: () => any`.
I think the two alternatives worthy to discuss are `onPress: () => any` and `onPress: () => void`. The first allows constructs like 
```
            onPress={() => dispatch(navigateToAddNewAccount(''))}
```
The second requires this to be
```
            onPress={() => {
              dispatch(navigateToAddNewAccount(''));
            }}
```
I prefer the second option, since it makes it clear that it's the content of the function that matters, and not the return type.

* `DiagnosticsScreen` seems to contain dead code. It has a prop `experimentalFeaturesEnabled` that's  used nowhere except in its function `handleExperimentalChange`, which isn't used anywhere. It looks like we should either remove this code, or finish the feature that was intended. Currently, this throws a type error because `experimentalFeaturesEnabled` is not specified when the component is created.

* `isStreamNarrow(narrow: Narrow)` and potentially other functions in narrow.js are sometimes called with `undefined`, even though `narrow: Narrow` isn't a maybe type. I noticed this when uncommenting the "FlowFixMe"s in Chat.js. I'm not sure if the issue here is that narrow is undefined or that we don't use a maybe type.

borisyankov commented:
You can import actions from the `actions.js` file which as
`import { someAction, anotherAction } form '../actions';`
thus not having multiple imports for multiple action files.

roberthoenig commented:
Added a commit for just importing `../actions` instead of individual action files. Also, I remembered one other observation I made yesterday:

The pattern of changes in these commits is mostly  straightforward: Use `connect` instead of `connectWithActions` and then just directly import the actions you wanna use. This works fine if the file is self-contained, that is, when it doesn't pass down the actions to other files. When it does, however, there are two scenarios:
* The first scenario is the https://github.com/zulip/zulip-mobile/pull/2648/commits/62d6d6db65a0e0bd95b2d7d7aa097c890270893f . Here, actions are passed down from AppEventHandlers to `notifications.*.js`. After changing `connectWithActions` to `connect`, we cannot pass down the actions anymore. In this case, this is fine because `notifications.*.js` is a "dead-end". It doesn't further pass down actions and we can just directly import actions in it as well. 

* The second scenario is a file like `MessageList.js`. I tried changing it from `connectWithActions` to `connect`, without success. This is because `MessageList.js` passes down its props (and with them its actinos) down a deep code path `executeActionSheetAction`, which then spreads out and sends the actions to further places, where they'll be eventually used. It'd be possible to resolve this with the same pattern as in the first scenario, but it'd require many more files to be changed. I'm not sure if we want to follow that code pattern, so asking here.

borisyankov commented:
@roberthoenig I think it is fine to leave the less straightforward pieces for later refactor. The ones you are pointing too are not as trivial indeed.

gnprice commented:
Thanks @roberthoenig ! Just merged all of these that don't contain `$FlowFixMe`; we don't have to fix the type errors immediately (as they were presumably always there, we're just exposing them now), but I'd like to have a few words of explanation in any "fixme" comments we merge.

I did go ahead and rebase your prettier fixes, and consolidation of imports, into the respective commits. The key tool that I find makes that fairly easy is to commit each affected file as a separate, temporary commit (with a terse commit message that's just enough to understand for the next minute or two -- a typical command I just ran is `git ci src/diagnostics/ -m 'prettier diagn screen'`, where `git ci` is an alias for `commit`); then `git rebase -i` to sort them to the right spot, and `f`/`fixup` there to fix up the respective commits with no further ado.

I pushed the rebase-squashed version of the whole branch to your PR branch, including the commits I didn't yet merge; so you might find it handy to pick up from there.

(I'll reply to your detailed comments and questions in a separate comment.)

gnprice commented:
> onPress: () => void

I agree this seems best given the function's return value will be ignored. Thanks for making that study of the RN docs!


> DiagnosticsScreen seems to contain dead code. It has a prop experimentalFeaturesEnabled that's used nowhere

Looks like this was last used to control the webview message list; and it was replaced for that purpose with 50cd02f, last October. So let's just cut this. Good catch!

I'd want to separate the commit that cuts it from the one that does the `dispatch` refactor; they can go in either order.


> isStreamNarrow(narrow: Narrow) and potentially other functions in narrow.js are sometimes called with undefined, even though narrow: Narrow isn't a maybe type.

Yeah -- my pilot commit for this refactor ran into a case of this too, as you can see in `Chat.js`. The conclusion I wrote in that comment is that we probably really do need a real `Narrow` there, and not undefined. Investigation needed, though.

gnprice commented:
> The first scenario is the 62d6d6d . Here, actions are passed down from AppEventHandlers to notifications.*.js. After changing connectWithActions to connect, we cannot pass down the actions anymore. In this case, this is fine because notifications.*.js is a "dead-end". It doesn't further pass down actions and we can just directly import actions in it as well.

Yeah, I'm pretty happy with this one. (Already merged it, in fact.)


> MessageList.js passes down its props (and with them its actinos) down a deep code path executeActionSheetAction, which then spreads out and sends the actions to further places, where they'll be eventually used.

Hmm, interesting. Looking at that code, I don't think I'm a fan of that arrangement -- passing the whole `props` down through several layers like that feels wrong, like the abstractions that are being used don't reflect what the code is really doing. That's probably not something to fix within this PR, though.

As a matter of strategy, I'll echo Boris that it's a fine plan to just leave this one for later. Some kinds of refactors depend on being carried out everywhere before they're useful, and for those it's important to do (or at least fully understand) the hard cases early; but this one is helpful independently for each place we do it, so it works great to sweep through the easier cases first and come back to the few hard ones at the end.

Looking just a bit closer at this code, I *think* -- I haven't confirmed thoroughly -- all those functions that `executeActionSheetAction` passes the `props` to are in that same file `messageActionSheet.js`. And just a few of them actually use `actions`, even. So it may end up being not that much work to go and replace `actions` with `dispatch` throughout, just as you've been doing elsewhere.

roberthoenig commented:
Updated:
* Added comments for `FlowFixMe`s
* Added commit to remove dead code in DiagnosticsScreen.js

Some of the comments are quite long I figure this shouldn't be a big problem, however, since we intend to fix these type errors timely?

roberthoenig commented:
Updated:
* changed all remaining easy-to-chagnge connectWithAction calls.
* Added commit to use `connect()` over `connect(null)` consistently.

Remaining calls to `connectWithActions`:
* `connectWithActions.js`. Called by connectWithActionsPreserveOnBack. Once we remove the other connectWithAction calls, we can remove this one.

*  `MessageList.js`. Passes `actions` down to other files, see comments above. Feasible to change, will just require some more work.

* `SearchMessagesCard.js`. Passes `actions` down to MessageList.js. Can change once MessageList is changed.

* 6 calls to `connectWithActionsPreserverOnBack`. `connectWithActionsPreserverOnBack` doesn't do any magic; It just calls `connectWithActions` with an additional mapDispatchToProps argument `{ areStatesEqual: isStateGoingBack }` that makes state changes more efficient. We can get rid of it just like with `connectWithActions`, except that we need to add `{ areStatesEqual: isStateGoingBack }` to every connect call.

borisyankov commented on src/account-info/AccountDetails.js:
```diff
@@ -51,6 +51,11 @@ export default class AccountDetails extends PureComponent<Props, void> {
+            we do not specify here. However, we don't need to, since the
+            `presence` prop gets automatically assigned in ActivityText's
+            creation with `connect`.
+            $FlowFixMe */}
```
You can mark the `presences` as optional ala:
```js
presences?: PresenceState,
```

Or maybe there is a better way to tell Flow the component is always getting the param from `connect`?
I think we should use $FlowFixMe sparingly, and possibly remove many of the ones we already have.

roberthoenig replied:
> I think we should use $FlowFixMe sparingly, and possibly remove many of the ones we already have.

I agree. This series of commits doesn't actually introduce any new Flow errors, it just reveals old ones that weren't noticed by Flow because of the code structure.

roberthoenig replied:
> You can mark the presences as optional ala: `presences?: PresenceState`

I don't think we should do this since it'd be semantically incorrect: `presences` is not optional, it is always required. Furthermore, we always satisfy this requirement in `connect`, it's just not recognized by Flow. 

> Or maybe there is a better way to tell Flow the component is always getting the param from connect?

That'd be great. I think as long as we haven't found such a better way, we should leave this FlowFixMe as it is. (and perhaps open an issue in Flow / react-redux once we are certain this is an error on their side?

borisyankov replied:
> presences is not optional, it is always required. 
From the outside this property is optional (not even needed).

roberthoenig replied:
I was thinking in terms of "`presence` will exist in every `ActivityText` object, therefore it shouldn't be marked optional in its Props". Otherwise, someone studying ActivityText.js  could wonder in what case `presences` would be optional, when this is actually never the case.

gnprice replied:
Yeah, there's really two "props" types here: the type `this.props` will have inside the main component (we conventionally call this `Props`), and the type of the props that a caller needs to pass to the container which is the default export (we don't generally write this down explicitly).

Here, clearly `presences` is part of one of these types but not the other.

The thing that puzzles me looking at this is: it's very common for us to have some properties that are in the inner type but not the outer one! There's a difference here basically every time we use `connect` (or, previously, `connectWithActions`); sometimes we take the property from the caller if present, but even then that means the type of that prop should be optional on the outside and not on the inside. So: why aren't we running into this more often?

gnprice replied:
Aha, I see -- it's that in that `connect` call, we claim the `props` parameter has type `Props`:
```
export default connect((state, props: Props) => ({
  presences: getPresence(state),
}))(ActivityText);
```

That's not right; but if we say so, then Flow correctly infers that the props passed to the container by its caller must have that type, and gives this error.

If I delete that annotation, the error goes away -- and also Flow successfully performs inference through it, so if I play with an `ActivityText` call site and introduce type errors, I get appropriate error messages.

So I'll just delete that annotation. It looks like that's the only place we make this mistake.

borisyankov replied:
We are using `?` more often :)

If we want to be both correct and precise, we would split the component into a `container` and a `dumb` component, then the container will not have this prop at all, and the `dumb` component will have this as `required` but it will also be clear that the component is passing props to it.

Another thing, in some Flow discussions people declare `own props` and `props from connect` as two different objects and then create a union between them.

Both of these are an improvement in but possibly too much boilerplate (or not?)

gnprice commented:
Thanks @roberthoenig ! Just merged all of these except the one on ActivityText, which I want to look at a bit closer related to the review thread above. Excited to see this nearly done :-D

Long comments are totally fine. In one commit, I did edit the comments a bit -- they ended up shorter, but the main thing I wanted to change was just to get the explanation right there on the same line as the `$FlowFixMe`:
```
-  // We have two Flow errors here because we are accessing props.narrow yet
-  // Props has no `narrow` property.
-  // Ignore the next line because prettier messes up the ternary below such
-  // that the linter complains.
+  // Prettier messes up the ternary so the linter complains.
   // prettier-ignore
   backgroundColor: !props.backgroundColor
-    // $FlowFixMe
+    // $FlowFixMe: Props has no `narrow` property
     ? getTitleBackgroundColor(props.narrow)(state)
     : props.backgroundColor,
-  // $FlowFixMe
+  // $FlowFixMe: Props has no `narrow` property
   textColor: getTitleTextColor(props.narrow)(state),
```

Where the explanation is longer, that's fine too; just want it as immediately next to the issue as possible.

(Sadly Prettier seems to get confused if the `prettier-ignore` comment adds any explanation at all -- I tried a couple of ways.)

gnprice commented:
Merged! I added a commit on top with the resolution to that puzzling `presences` type issue. Thanks again.

roberthoenig commented:
Cool, closing this then. WIll open a new PR for the remaining cases.

roberthoenig commented:
Huh, just noticed there is no close button ^^. Is that because it has been merged?
